tests/test_wordpress_api.py

```diff
@@ -116,157 +30,162 @@ def test_getting_all_articles(self, get):
             + "&exclude="
             + "&author="
         )
-        self.assertEqual(article, (["hello_test"], 12))
 
-    @patch("canonicalwebteam.http.CachedSession.get")
-    def test_excluding_articles(self, get):
-
-        get.return_value = MockResponse()
-
-        article = api.get_articles(tags_exclude=[1234, 5678])
-        get.assert_called_once_with(
-            "https://admin.insights.ubuntu.com/"
-            + "wp-json/wp/v2/posts?"
-            + "per_page=12"
-            + "&tags="
-            + "&page=1"
-            + "&group="
-            + "&tags_exclude=1234,5678"
-            + "&categories="
-            + "&exclude="
-            + "&author="
+        url = api.build_get_articles_url(page=3, per_page=24)
+        self.assertEqual(
```

Comment:
I think this test is getting a little too specific - checking the exact format of the built URL certainly feels like testing internals.

Personally I'd suggest you treat `build_get_articles_url` as an internal function / private method of `api` (maybe rename it to `_build_get_articles_url` [to signify this](https://docs.python.org/3/tutorial/classes.html#private-variables)) and so say that it shouldn't be tested at all. Instead, I think it would be ideal to only test `get_articles` and `get_articles_with_metadata` - as these form the important external interface of the API library - and just make sure they return the expected articles. Of course this means you have to have fixtures for all the different filtered API responses.

Reply:
Good idea! I think its worth the effort, as this is such a core functionality

Reply:
So, I have left this unittest in for now, as the function itself needs updating. 
No harm in having an extra unit-test, no?

Reply:
There is harm. The harm is that if some well-meaning soul wants to refactor the code in a way that doesn't effect the actual external interface of the module, their job will be that much harder. They have to either feel empowered to delete this test, or attempt to rewrite it.

But even if you were going to leave it in place, I would say it's better to do a test something like:

``` python3
self.assertTrue('page=3' in urlparse(url).query)
```

rather than testing the exact format of the URL output. Then at least you're accommodating changes in the order of the parameters, will have zero effect on the functionality.

Reply:
While I agree in general, I think this function has a lot of responsibility and should be guarded.

I do like the `assertTrue` logic a lot better though, it makes it clearer what the test does. I will update it now :)

Reply:
I still want to defend the principle that in a well-written module, only the external interface should need to be tested. This is because a neat module won't have gold-plated functions, each function will only do as much as is needed by the actual functions of the module. Therefore, it should be possible to test every line of code in the module by simply testing each of the module's external functions. If this isn't possible, code should be removed from the internals.

In very complex modules, I can see that there might be cases where trying to test an internal line of code from the outside might require a very convoluted test, which might then be very hard to maintain. In this case there could be an argument for testing an internal function so as to keep tests simpler and maintainable.

But in this case I don't think this argument holds. Everything you're testing with `build_get_articles_url` could trivially be instead tested with `get_articles` or `get_articles_with_metadata` - and in fact they really should be. The really important thing to test here is that `get_articles(page=3)` works, because that is the actual interface that clients will use, which you're not actually testing.

Reply:
Click! Got it now and updated the tests